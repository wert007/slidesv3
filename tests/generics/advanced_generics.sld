struct A<Type> {
    value: $Type;

    func print() {
        print(this.value);
    }

    func printBoxed() {
        boxedPrint(this.value);
    }

    func printGreeted() {
        print('Hello:');
        this.print();
    }
}

func boxedPrint(msg: any) {
    print('BOX: ' + msg);
}

func main() {
    let intVersion = new A(3);
    let boolVersion = new A(false);
    let stringVersion = new A('Hello');
    intVersion.print();
    boolVersion.print();
    stringVersion.print();
    intVersion.printBoxed();
    boolVersion.printBoxed();
    stringVersion.printBoxed();
    intVersion.printGreeted();
    boolVersion.printGreeted();
    stringVersion.printGreeted();

    let a = new TwoLevelGeneric(1, [1, 2, 1, 1]);
    a.check();
}

struct TwoLevelGeneric<Type> {
    expected_mean: $Type;
    values: $Type[];

    func check() {
        for index, value in this.values {
            if value == this.expected_mean {
                print('Index ' + index + ' passed check!');
            }
        }
    }
}
