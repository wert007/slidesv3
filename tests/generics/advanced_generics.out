3
false
Hello
BOX: 3
BOX: false
BOX: Hello
Hello:
3
Hello:
false
Hello:
Hello
0: 0
1: 0
2: 0
3: 0
4: 0
5: 0
= -1
